ptypy/engines/ML.py

```diff
@@ -290,6 +315,13 @@ def engine_iterate(self, num=1):
             self.pr_grad *= self.scale_p_o
             self.pr_h -= self.pr_grad
 
+            # Lipschitz preconditioner
```

Comment:
As just discussed, this application of the preconditioner needs to be moved up before the first application of the smoothing preconditioner (line 257), and needs to be applied once more in a way similar to what is done at lines 308-312.

Reply:
Unfortunately I can't make it work when the preconditioner is applied twice as suggested (see the commented out code), I get the following:
![ml_precond_twice](https://github.com/user-attachments/assets/7f1768a2-845a-4ae9-80dc-e1ed141aaf34)
Are we missing something fundamental here? I think of this more like a damping parameter as in the paper:

> The weightings can be seen as preconditioners of the gradient descent task [19, 21] with small regularization constants δP, δO > 0. These regularized preconditioners result in a behavior similar to the damped least-squares method that penalizes large values in the update, particularly in the weakly illuminated regions.

Reply:
I tried applying the sqrt of the inverse fluence map twice, but the results are still not as good as before (see the notebook):
![ML_wavefield](https://github.com/user-attachments/assets/de24f3f7-769e-4a18-ac7c-f78e39200927)
Something is still not quite right in the code somewhere...

Reply:
Done and working as expected.